src/data_ingestion/readers/data_reader_python_dataset.cpp

```diff
@@ -0,0 +1,481 @@
+  const std::string sample_func_name =
+    ("_DATA_READER_PYTHON_CPP_sample_function_wrapper" +
+     std::to_string(instance_id));
+  PyObject_SetAttrString(main_module,
+                         sample_func_name.c_str(),
+                         m_sample_function);
+  python::check_error();
+  const std::string shared_array_name =
+    ("_DATA_READER_PYTHON_CPP_shared_memory_array" +
+     std::to_string(instance_id));
+  PyObject_SetAttrString(main_module,
+                         shared_array_name.c_str(),
+                         m_shared_memory_array);
+  python::check_error();
+
+  std::string label_shared_array_name = "None";
+  if (has_labels()) {
+    m_label_shared_memory_array =
+      PyObject_CallMethod(multiprocessing_module,
+                          "RawArray",
+                          "(s, l)",
+                          datatype_typecode.c_str(),
+                          get_num_labels() * mini_batch_size);
+    python::object label_shared_memory_ptr =
+      PyObject_CallMethod(ctypes_module,
+                          "addressof",
+                          "(O)",
+                          m_label_shared_memory_array.get());
+    m_label_shared_memory_array_ptr =
+      reinterpret_cast<DataType*>(PyLong_AsLong(label_shared_memory_ptr));
+    label_shared_array_name =
+      ("_DATA_READER_PYTHON_CPP_label_shared_memory_array" +
+       std::to_string(instance_id));
+    PyObject_SetAttrString(main_module,
+                           label_shared_array_name.c_str(),
+                           m_label_shared_memory_array);
+    python::check_error();
+  }
+
+  std::string response_shared_array_name = "None";
+  if (has_responses()) {
+    m_response_shared_memory_array =
+      PyObject_CallMethod(multiprocessing_module,
+                          "RawArray",
+                          "(s, l)",
+                          datatype_typecode.c_str(),
+                          get_num_responses() * mini_batch_size);
+    python::object response_shared_memory_ptr =
+      PyObject_CallMethod(ctypes_module,
+                          "addressof",
+                          "(O)",
+                          m_response_shared_memory_array.get());
+    m_response_shared_memory_array_ptr =
+      reinterpret_cast<DataType*>(PyLong_AsLong(response_shared_memory_ptr));
+    response_shared_array_name =
+      ("_DATA_READER_PYTHON_CPP_response_shared_memory_array" +
+       std::to_string(instance_id));
+    PyObject_SetAttrString(main_module,
+                           response_shared_array_name.c_str(),
+                           m_response_shared_memory_array);
+    python::check_error();
+  }
+
+  /// @todo Implement label reconstruction case
+
+  // Create wrapper around sample function
+  // Note: We attempt accessing the sample with the buffer protocol
+  // since they can be copied more efficiently. If this fails, we just
+  // iterate through the sample entries.
+  /// @todo Handle multi-dimensional NumPy arrays.
```

Comment:
it should be fine. `memoryview`, iirc, creates a flat buffer, and then you reshape that in LBANN. What it won't support is noncontiguous buffers (strided).

Reply:
pybind11 has a nice interface to handle numpy arrays

Reply:
No longer using memoryviews